packages/instrumentation-amqplib/test/amqplib-promise.spec.ts

```diff
@@ -0,0 +1,514 @@
+            expect(memoryExporter.getFinishedSpans()[1].status.code).toStrictEqual(SpanStatusCode.ERROR);
+            expect(memoryExporter.getFinishedSpans()[1].status.message).toStrictEqual(
+                'reject called on message with requeue'
+            );
+            expect(memoryExporter.getFinishedSpans()[2].status.code).toStrictEqual(SpanStatusCode.ERROR);
+            expect(memoryExporter.getFinishedSpans()[2].status.message).toStrictEqual(
+                'reject called on message without requeue'
+            );
+            expectConsumeEndSpyStatus([EndOperation.Reject, EndOperation.Reject]);
+        });
+
+        it('closing channel should end all open spans on it', async () => {
+            lodash.times(1, () => channel.sendToQueue(queueName, Buffer.from(msgPayload)));
+
+            await new Promise<void>((resolve) =>
+                asyncConsume(channel, queueName, [
+                    async (msg) => {
+                        await channel.close();
+                        resolve();
+                    },
+                ])
+            );
+
+            expect(memoryExporter.getFinishedSpans().length).toBe(2);
+            expect(memoryExporter.getFinishedSpans()[1].status.code).toStrictEqual(SpanStatusCode.ERROR);
+            expect(memoryExporter.getFinishedSpans()[1].status.message).toStrictEqual('channel closed');
+            expectConsumeEndSpyStatus([EndOperation.ChannelClosed]);
+        });
+
+        it('error on channel should end all open spans on it', (done) => {
+            lodash.times(2, () => channel.sendToQueue(queueName, Buffer.from(msgPayload)));
+
+            channel.on('close', () => {
+                expect(memoryExporter.getFinishedSpans().length).toBe(4);
+                // second consume ended with valid ack, previous message not acked when channel is errored.
+                // since we first ack the second message, it appear first in the finished spans array
+                expect(memoryExporter.getFinishedSpans()[2].status.code).toStrictEqual(SpanStatusCode.UNSET);
+                expect(memoryExporter.getFinishedSpans()[3].status.code).toStrictEqual(SpanStatusCode.ERROR);
+                expect(memoryExporter.getFinishedSpans()[3].status.message).toStrictEqual('channel error');
+                expectConsumeEndSpyStatus([EndOperation.Ack, EndOperation.ChannelError]);
+                done();
+            });
+            asyncConsume(channel, queueName, [
+                null,
+                (msg) => {
+                    try {
+                        channel.ack(msg);
+                        // ack the same msg again, this is not valid and should close the channel
+                        channel.ack(msg);
+                    } catch {}
+                },
+            ]);
+        });
+
+        // what should we do in this case?
+        // can cause memory leak since the plugin saves a copy of the msg
```

Comment:
Is this a real issue that can happen in production?

Reply:
This is a rare issue that could happen, in which case plugin will hold a pointer to the msg, which is no longer held by the user, but will never be "foggoten" by the plugin for garbage collection.

This was my progress so far:
Exception can happen in addition with `ack` / `nack` / `reject` / `ackAll` / `nackAll`, or on it's own:
- can `throw` before acknowledgment call which will never arrive
- can first acknoladge the message which will end the span, and then `throw` exception which will not be documented in the trace since it arrive after span ended.
- can dispatch async function execution which will ack sometime in the future, after exception is thrown from the original context.

I should deeply think and test the various possibilities and how plugin will react in each one to make sure we are not crashing or producing an inconsistent experience (or at least document this behavior somewhere).

Thanks for bringing it up

Reply:
Another option is to use [`WeakSet`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet) which offers a guaranty not to leak memory in any situation. This is a common practice in many instrumentation libraries, for example [instrumentation http](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-instrumentation-http/src/http.ts#L61) and a couple of web instrumentations.

In this case, since I need to know the sequence of messages to properly handle [`{n}ackAll`](https://www.squaremobius.net/amqp.node/channel_api.html#channel_ackAll) and the [`allUpTo`](https://www.squaremobius.net/amqp.node/channel_api.html#channel_ack) option, I need something like "weakArray", or a way to enumerate the weak collection which is [not possible by design](https://stackoverflow.com/questions/36760314/how-to-make-weak-set-or-weak-map-iterable-in-es6).

Would love to discuss this issue and think of the best approach